src/goto-analyzer/static_simplifier.cpp

```diff
@@ -0,0 +1,307 @@
+{
+  exprt simplified = domain[i_it].domain_simplify(i_it->guard, ns);
+  unsigned return_value = (simplified == i_it->guard) ? 0 : 1;
+  i_it->guard = simplified;
+  return return_value;
+}
+
+/*******************************************************************\
+
+Function: static_simplifiert<analyzerT>::simplify_program
+
+  Inputs: None.
+
+ Outputs: None.
+
+ Purpose: Simplifies the program using the information in the abstract domain.
+
+\*******************************************************************/
+
+template <class analyzerT>
+void static_simplifiert<analyzerT>::simplify_program()
+{
+  struct counters {
+    unsigned asserts;
+    unsigned assumes;
+    unsigned gotos;
+    unsigned assigns;
+    unsigned function_calls;
+  };
+  
+  counters simplified = {0,0,0,0,0};
+  counters unmodified = {0,0,0,0,0};
+
+  Forall_goto_functions(f_it, goto_functions)
+  {
+    Forall_goto_program_instructions(i_it, f_it->second.body)
+    {
+      if (i_it->is_assert())
+      {
+	unsigned result = simplify_guard(i_it);
+	simplified.asserts += result;
+	unmodified.asserts += (1 - result);
+      }
+      else if (i_it->is_assume())
+      {
+	unsigned result = simplify_guard(i_it);
+	simplified.assumes += result;
+	unmodified.assumes += (1 - result);
+      }
+      else if (i_it->is_goto())
+      {
+	unsigned result = simplify_guard(i_it);
+	simplified.gotos += result;
+	unmodified.gotos += (1 - result);
+      }
+      else if (i_it->is_assign())
+      {
+	code_assignt assign(to_code_assign(i_it->code));
+
+	/*
+	** Simplification needs to be aware of which side of the
+	** expression it is handling as:
+	**   <i = 0, j = 1>  i = j
+	** should simplify to i = 1, not to 0 = 1.
+	*/
+	
+	exprt simplified_lhs = domain[i_it].domain_simplify(assign.lhs(), ns, true);
+	exprt simplified_rhs = domain[i_it].domain_simplify(assign.rhs(), ns, false);
+
+	unsigned result = (simplified_lhs == assign.lhs() &&
+			   simplified_rhs == assign.rhs()) ? 0 : 1;
+	simplified.assigns += result;
+	unmodified.assigns += (1 - result);
+
+	assign.lhs() = simplified_lhs;
+	assign.rhs() = simplified_rhs;
+	i_it->code = assign;
+      }
+      else if (i_it->is_function_call())
+      {
+	unsigned result = 0;
+	code_function_callt fcall(to_code_function_call(i_it->code));
+	
+	exprt new_function = domain[i_it].domain_simplify(fcall.function(), ns);
+	result |= (new_function == fcall.function()) ? 0 : 1;
+	fcall.function() = new_function;
+	
+	exprt::operandst &args = fcall.arguments();
+
+	for(exprt::operandst::iterator o_it = args.begin();
+          o_it != args.end(); ++o_it)
+	{
+	  exprt new_arg = domain[i_it].domain_simplify(*o_it, ns);
+	  result |= (new_arg == *o_it) ? 0 : 1;
+	  *o_it = new_arg;
+	}
+
+	i_it->code = fcall;
+	
+	simplified.function_calls += result;
+	unmodified.function_calls += (1 - result);
+      }
+    }
+  }
+  
+  //make sure the references are correct
+  goto_functions.update();
+
+    
+  status() << "SIMPLIFIED: "
```

Comment:
Is this status or statistics? Is this output ui-dependent (just to make sure that no-one expects this to work with xml or json)?